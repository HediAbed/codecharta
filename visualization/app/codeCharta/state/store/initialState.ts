import { State } from "../../codeCharta.model"
import { Vector3 } from "three"

// This will be gone, once all the initialStates are specific on their reducer
export const initialState: State = {
	fileSettings: {
		attributeTypes: { nodes: [], edges: [] },
		blacklist: [],
		edges: [],
		markedPackages: []
	},
	dynamicSettings: {
		areaMetric: null,
		heightMetric: null,
		colorMetric: null,
		distributionMetric: null,
		edgeMetric: null,
		focusedNodePath: "",
		searchedNodePaths: [],
		searchPattern: "",
		margin: null,
		colorRange: { from: null, to: null }
	},
	appSettings: {
		amountOfTopLabels: 1,
		amountOfEdgePreviews: 1,
		edgeHeight: 4,
		scaling: new Vector3(1, 1, 1),
		camera: new Vector3(0, 300, 1000),
		hideFlatBuildings: true,
		invertColorRange: false,
		invertDeltaColors: false,
		invertHeight: false,
		dynamicMargin: true,
		isWhiteBackground: false,
		whiteColorBuildings: false,
		mapColors: {
			positive: "#69AE40",
			neutral: "#ddcc00",
			negative: "#820E0E",
			selected: "#EB8319",
			defaultC: "#89ACB4",
			positiveDelta: "#69FF40",
			negativeDelta: "#ff0E0E",
			base: "#666666",
			flat: "#AAAAAA",
			lightGrey: "#DDDDDD",
			angularGreen: "#00BFA5",
			markingColors: ["#FF1D8E", "#1d8eff", "#1DFFFF", "#8eff1d", "#8e1dff", "#FFFF1D"],
			incomingEdge: "#00ffff",
			outgoingEdge: "#ff00ff"
		},
		isPresentationMode: false,
		showOnlyBuildingsWithEdges: false,
		resetCameraIfNewFileIsLoaded: true
	},
	treeMapSettings: {
		mapSize: 250
	}
}
