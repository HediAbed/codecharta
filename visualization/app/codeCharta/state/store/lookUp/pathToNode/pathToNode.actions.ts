import { Action } from "redux"
import { CodeMapNode } from "../../../../codeCharta.model"

export enum PathToNodeActions {
	SET_PATH_TO_NODE = "SET_PATH_TO_NODE"
}

export interface SetPathToNodeAction extends Action {
	type: PathToNodeActions.SET_PATH_TO_NODE
	payload: Map<string, CodeMapNode>
}

export type PathToNodeAction = SetPathToNodeAction

export function setPathToNode(pathToNode: Map<string, CodeMapNode> = defaultPathToNode): SetPathToNodeAction {
	return {
		type: PathToNodeActions.SET_PATH_TO_NODE,
		payload: pathToNode
	}
}

export const defaultPathToNode: Map<string, CodeMapNode> = new Map()
