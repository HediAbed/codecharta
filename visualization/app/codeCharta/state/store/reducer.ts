import appSettings from "./appSettings/appSettings.reducer"
import fileSettings from "./fileSettings/fileSettings.reducer"
import dynamicSettings from "./dynamicSettings/dynamicSettings.reducer"
import { CCAction, State } from "../../codeCharta.model"
import { StateActions } from "./state.actions"
import { setFileSettings } from "./fileSettings/fileSettings.actions"
import { setAppSettings } from "./appSettings/appSettings.actions"
import { setDynamicSettings } from "./dynamicSettings/dynamicSettings.actions"

export function rootReducer(state: State = {} as State, action: CCAction) {
	let fileSettingsAction = action
	let appSettingsAction = action
	let dynamicSettingsAction = action

	if (action.type === StateActions.SET_STATE) {
		if (action.payload.fileSettings) {
			fileSettingsAction = setFileSettings(action.payload.fileSettings)
		}

		if (action.payload.appSettings) {
			appSettingsAction = setAppSettings(action.payload.appSettings)
		}

		if (action.payload.dynamicSettings) {
			dynamicSettingsAction = setDynamicSettings(action.payload.dynamicSettings)
		}
	}

	return {
		fileSettings: fileSettings(state.fileSettings, fileSettingsAction),
		appSettings: appSettings(state.appSettings, appSettingsAction),
		dynamicSettings: dynamicSettings(state.dynamicSettings, dynamicSettingsAction)
	}
}
