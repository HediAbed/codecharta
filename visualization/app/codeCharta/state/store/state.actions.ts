import { CCAction, RecursivePartial, State } from "../../codeCharta.model"

export enum StateActions {
	SET_STATE = "SET_STATE"
}

export interface SetStateAction extends CCAction {
	type: StateActions.SET_STATE
	payload: RecursivePartial<State>
}

export type StateAction = SetStateAction

export function setState(state: RecursivePartial<State>): StateAction {
	return {
		type: StateActions.SET_STATE,
		payload: state
	}
}
