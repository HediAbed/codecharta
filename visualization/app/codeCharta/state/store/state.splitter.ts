import _ from "lodash"
import { CCAction } from "../../codeCharta.model"
import { StateActions } from "./state.actions"
import { DynamicSettingsActions } from "./dynamicSettings/dynamicSettings.actions"
import { FileSettingsActions } from "./fileSettings/fileSettings.actions"
import { AppSettingsActions } from "./appSettings/appSettings.actions"
import { splitDynamicSettingsActions } from "./dynamicSettings/dynamicSettings.splitter"
import { splitFileSettingsActions } from "./fileSettings/fileSettings.splitter"
import { splitAppSettingsActions } from "./appSettings/appSettings.splitter"

export function splitInAtomicActions(action: CCAction): CCAction[] {
	if (_.values(DynamicSettingsActions).includes(action.type)) {
		return splitDynamicSettingsActions(action.payload.dynamicSettings)
	}

	if (_.values(FileSettingsActions).includes(action.type)) {
		return splitFileSettingsActions(action.payload.fileSettings)
	}

	if (_.values(AppSettingsActions).includes(action.type)) {
		return splitAppSettingsActions(action.payload.appSettings)
	}

	if (_.values(StateActions).includes(action.type)) {
		const actions: CCAction[] = []

		if (action.payload.dynamicSettings !== undefined) {
			actions.concat(splitDynamicSettingsActions(action.payload.dynamicSettings))
		}

		if (action.payload.fileSettings !== undefined) {
			actions.concat(splitFileSettingsActions(action.payload.fileSettings))
		}

		if (action.payload.appSettings !== undefined) {
			actions.concat(splitAppSettingsActions(action.payload.appSettings))
		}
		return actions
	}

	return [action]
}
