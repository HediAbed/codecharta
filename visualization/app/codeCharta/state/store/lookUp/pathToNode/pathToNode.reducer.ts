import { PathToNodeAction, PathToNodeActions, setPathToNode } from "./pathToNode.actions"
import { CodeMapNode } from "../../../../codeCharta.model"

export function pathToNode(state: Map<string, CodeMapNode> = setPathToNode().payload, action: PathToNodeAction): Map<string, CodeMapNode> {
	switch (action.type) {
		case PathToNodeActions.SET_PATH_TO_NODE:
			return new Map(action.payload)
		default:
			return state
	}
}
