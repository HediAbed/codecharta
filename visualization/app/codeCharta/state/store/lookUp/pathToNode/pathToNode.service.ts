import { StoreService, StoreSubscriber } from "../../../store.service"
import { IRootScopeService } from "angular"
import { PathToNodeActions, setPathToNode } from "./pathToNode.actions"
import { isActionOfType } from "../../../../util/reduxHelper"
import { CodeMapNode } from "../../../../codeCharta.model"
import { CodeMapPreRenderService, CodeMapPreRenderServiceSubscriber } from "../../../../ui/codeMap/codeMap.preRender.service"
import { hierarchy } from "d3"

export interface PathToNodeSubscriber {
	onPathToNodeChanged(pathToNode: Map<string, CodeMapNode>)
}

export class PathToNodeService implements StoreSubscriber, CodeMapPreRenderServiceSubscriber {
	private static PATH_TO_NODE_CHANGED_EVENT = "path-to-node-changed"

	constructor(private $rootScope: IRootScopeService, private storeService: StoreService) {
		StoreService.subscribe(this.$rootScope, this)
		CodeMapPreRenderService.subscribe(this.$rootScope, this)
	}

	public onStoreChanged(actionType: string) {
		if (isActionOfType(actionType, PathToNodeActions)) {
			this.notify(this.select())
		}
	}

	public onRenderMapChanged(map: CodeMapNode) {
		const pathToNode = new Map<string, CodeMapNode>()
		pathToNode.set(map.path, map)
		hierarchy(map)
			.leaves()
			.forEach(x => {
				pathToNode.set(x.data.path, x.data)
			})

		this.storeService.dispatch(setPathToNode(pathToNode))
	}

	private select() {
		return this.storeService.getState().lookUp.pathToNode
	}

	private notify(newState: Map<string, CodeMapNode>) {
		this.$rootScope.$broadcast(PathToNodeService.PATH_TO_NODE_CHANGED_EVENT, { pathToNode: newState })
	}

	public static subscribe($rootScope: IRootScopeService, subscriber: PathToNodeSubscriber) {
		$rootScope.$on(PathToNodeService.PATH_TO_NODE_CHANGED_EVENT, (event, data) => {
			subscriber.onPathToNodeChanged(data.pathToNode)
		})
	}
}
