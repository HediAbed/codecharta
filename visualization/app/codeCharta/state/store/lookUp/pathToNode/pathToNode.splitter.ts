import { PathToNodeAction, setPathToNode } from "./pathToNode.actions"
import { CodeMapNode } from "../../../../codeCharta.model"

export function splitPathToNodeAction(payload: Map<string, CodeMapNode>): PathToNodeAction {
	return setPathToNode(payload)
}
