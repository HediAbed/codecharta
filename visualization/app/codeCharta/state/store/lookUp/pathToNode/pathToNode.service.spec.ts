import "../../../state.module"
import { IRootScopeService } from "angular"
import { StoreService } from "../../../store.service"
import { getService, instantiateModule } from "../../../../../../mocks/ng.mockhelper"
import { PathToNodeAction, PathToNodeActions } from "./pathToNode.actions"
import { PathToNodeService } from "./pathToNode.service"
import { TEST_FILE_WITH_PATHS, withMockedEventMethods } from "../../../../util/dataMocks"
import { CodeMapNode } from "../../../../codeCharta.model"
import { CodeMapPreRenderService } from "../../../../ui/codeMap/codeMap.preRender.service"

describe("PathToNodeService", () => {
	let pathToNodeService: PathToNodeService
	let storeService: StoreService
	let $rootScope: IRootScopeService

	beforeEach(() => {
		restartSystem()
		rebuildService()
		withMockedEventMethods($rootScope)
	})

	function restartSystem() {
		instantiateModule("app.codeCharta.state")

		$rootScope = getService<IRootScopeService>("$rootScope")
		storeService = getService<StoreService>("storeService")
	}

	function rebuildService() {
		pathToNodeService = new PathToNodeService($rootScope, storeService)
	}

	describe("constructor", () => {
		it("should subscribe to store", () => {
			StoreService.subscribe = jest.fn()

			rebuildService()

			expect(StoreService.subscribe).toHaveBeenCalledWith($rootScope, pathToNodeService)
		})

		it("should subscribe to renderMapChange", () => {
			CodeMapPreRenderService.subscribe = jest.fn()

			rebuildService()

			expect(CodeMapPreRenderService.subscribe).toHaveBeenCalledWith($rootScope, pathToNodeService)
		})
	})

	describe("onStoreChanged", () => {
		it("should notify all subscribers with the new pathToNode value", () => {
			const map = new Map<string, CodeMapNode>()
			map.set(TEST_FILE_WITH_PATHS.map.path, TEST_FILE_WITH_PATHS.map)
			const action: PathToNodeAction = {
				type: PathToNodeActions.SET_PATH_TO_NODE,
				payload: map
			}
			storeService["store"].dispatch(action)

			pathToNodeService.onStoreChanged(PathToNodeActions.SET_PATH_TO_NODE)

			expect($rootScope.$broadcast).toHaveBeenCalledWith("path-to-node-changed", { pathToNode: map })
		})

		it("should not notify anything on non-path-to-node-events", () => {
			pathToNodeService.onStoreChanged("ANOTHER_ACTION")

			expect($rootScope.$broadcast).not.toHaveBeenCalled()
		})
	})

	describe("onRenderMapChanged", () => {
		it("should update the map", () => {
			pathToNodeService.onRenderMapChanged(TEST_FILE_WITH_PATHS.map)

			expect(storeService.getState().lookUp.pathToNode).toMatchSnapshot()
		})
	})
})
