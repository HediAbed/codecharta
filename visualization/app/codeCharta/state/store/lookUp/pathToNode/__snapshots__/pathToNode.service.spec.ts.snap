// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PathToNodeService onRenderMapChanged should update the map 1`] = `
Map {
  "/root" => Object {
    "attributes": Object {},
    "children": Array [
      Object {
        "attributes": Object {
          "functions": 10,
          "mcc": 1,
          "rloc": 100,
        },
        "link": "http://www.google.de",
        "name": "big leaf",
        "path": "/root/big leaf",
        "type": "File",
      },
      Object {
        "attributes": Object {},
        "children": Array [
          Object {
            "attributes": Object {
              "functions": 100,
              "mcc": 100,
              "rloc": 30,
            },
            "name": "small leaf",
            "path": "/root/Parent Leaf/small leaf",
            "type": "File",
          },
          Object {
            "attributes": Object {
              "functions": 1000,
              "mcc": 10,
              "rloc": 70,
            },
            "name": "other small leaf",
            "path": "/root/Parent Leaf/other small leaf",
            "type": "File",
          },
          Object {
            "attributes": Object {},
            "children": Array [],
            "name": "empty folder",
            "path": "/root/Parent Leaf/empty folder",
            "type": "Folder",
          },
        ],
        "name": "Parent Leaf",
        "path": "/root/Parent Leaf",
        "type": "Folder",
      },
    ],
    "name": "root",
    "path": "/root",
    "type": "Folder",
  },
  "/root/big leaf" => Object {
    "attributes": Object {
      "functions": 10,
      "mcc": 1,
      "rloc": 100,
    },
    "link": "http://www.google.de",
    "name": "big leaf",
    "path": "/root/big leaf",
    "type": "File",
  },
  "/root/Parent Leaf/small leaf" => Object {
    "attributes": Object {
      "functions": 100,
      "mcc": 100,
      "rloc": 30,
    },
    "name": "small leaf",
    "path": "/root/Parent Leaf/small leaf",
    "type": "File",
  },
  "/root/Parent Leaf/other small leaf" => Object {
    "attributes": Object {
      "functions": 1000,
      "mcc": 10,
      "rloc": 70,
    },
    "name": "other small leaf",
    "path": "/root/Parent Leaf/other small leaf",
    "type": "File",
  },
  "/root/Parent Leaf/empty folder" => Object {
    "attributes": Object {},
    "children": Array [],
    "name": "empty folder",
    "path": "/root/Parent Leaf/empty folder",
    "type": "Folder",
  },
}
`;
